### PR TITLE
[BRE-443] Add Project lock functionality and refactor code

### PR DESCRIPTION
**Pull request by ragurney**
screen shot 2018-06-25 at 4 52 07 pm
screen shot 2018-06-25 at 5 31 19 pm

/cc @zendesk/samson @zendesk/bre 

### References
 - Jira link: [BRE-443](https://zendesk.atlassian.net/browse/BRE-443)

### CODE REVIEW

File: app/models/access_control.rb [hunk 1]
```diff
@@ -18,12 +18,12 @@ def can?(user, action, resource_namespace, scope = nil)
...
-          return false unless scope.is_a?(Stage) # stage locks
-          if ENV['PRODUCTION_STAGE_LOCK_REQUIRES_ADMIN'] && scope.production
-            user.admin_for?(scope.project)
+          case scope.class.to_s
+          when 'NilClass' then user.admin? # global locks
+          when 'Project' then user.admin_for?(scope) # project locks
+          when 'Stage' then can_lock_stage?(user, scope) # stage locks
           else
-            user.deployer_for?(scope.project)
+            false
```
Comment (grosser) [inline app/models/access_control.rb]: for more symmetry could even do `else false`

File: app/models/access_control.rb [hunk 2]
```diff
@@ -65,5 +65,13 @@ def can?(user, action, resource_namespace, scope = nil)
     def can_deploy_anything?(user)
       user.deployer? || user.user_project_roles.where('role_id >= ?', Role::DEPLOYER.id).exists?
     end
+
+    def can_lock_stage?(user, stage)
+      if ENV['PRODUCTION_STAGE_LOCK_REQUIRES_ADMIN'] && stage.production
```
Comment (grosser) [inline app/models/access_control.rb]: I think `.production` != `.production?` (column vs logic)

Comment (grosser) [reply app/models/access_control.rb]: (pre-existing bug I guess, can still solve)

File: app/models/stage.rb [hunk 3]
```diff
@@ -204,6 +205,10 @@ def url
     Rails.application.routes.url_helpers.project_stage_url(project, self)
   end
 
+  def affected_by_lock?(lock)
```
Comment (grosser) [inline app/models/stage.rb]: you sneaky :D

File: app/models/stage.rb [hunk 4]
```diff
@@ -254,4 +259,12 @@ def append_new_command
...
     stage_commands.build(command: new_command, position: previous + 1)
   end
+
+  def environment_lock?(lock)
+    lock.resource_type == "Environment" && environments.any? { |e| lock.resource_equal(e) }
+  end
+
+  def project_lock?(lock)
+    lock.resource_type == "Project" && lock.resource_equal(project)
+  end
```
Comment (grosser) [inline app/models/stage.rb]: this feels kinda wrong ... locking logic is now all over the place :(

File: app/models/concerns/lockable.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+module Lockable
+  extend ActiveSupport::Concern
+
+  included do
+    has_one :lock, as: :resource
+  end
+
+  def affected_by_lock?(lock)
```
Comment (grosser) [inline app/models/concerns/lockable.rb]: `def locked_by?(lock)`

File: app/models/lock.rb [hunk 6]
```diff
@@ -23,8 +23,36 @@ class Lock < ActiveRecord::Base
...
+    # normally there are very few locks, so we grab them all and filter down to avoid lookups
+    # sorted by priority(warning) and specificity(type)
+    def for_resource(resource)
+      matching_locks = all_cached.select { |l| resource.affected_by_lock?(l) }
+      matching_locks.sort_by! { |l| [l.warning? ? 1 : 0, RESOURCE_TYPES.index(l.resource_type)] }
+    end
+
+    def locked_for?(resource, user)
+      locks = for_resource(resource)
+      locks.any? { |l| !l.warning? && l.user != user }
```
Comment (grosser) [inline app/models/lock.rb]: compare the ids to avoid lookups ?

File: app/models/access_control.rb (hunk 2 above)
Comment (ragurney) [reply app/models/access_control.rb]: Good catch! Will fix

File: app/models/stage.rb (hunk 4 above)
Comment (ragurney) [reply app/models/stage.rb]: Yeah... I was on the fence about this. It kind of smells bad, but as you say it allows for locks to be used in plugins and it moves model-specific logic to the models.

**Pull request merged**